### PR TITLE
gh-648 Provide infrastructure for local E2E/Unit Testing SauceLabs/Browserstack

### LINKED ISSUE #648: Investigate E2E JS testing services

### PR DESCRIPTION
**Pull request by ghillert**
- Update Readme
- Add support for SauceLabs-based E2E testing
- Add support for SauceLabs-based Unit testing
- Add support for BrowserStack-based E2E testing
- Add support for BrowserStack-based Unit testing

In order to run the tests, developers must provide **SauceLabs credentials**, e.g.:

```bash
    $ export SAUCE_USERNAME=your-username
    $ export SAUCE_ACCESS_KEY=your-access-key
```

and **BrowserStack credentials**, e.g.:

```bash
    $ export BROWSER_STACK_USERNAME=your-username
    $ export BROWSER_STACK_ACCESS_KEY=your-access-key
```

Important, for E2E tests, developers should refrain from using `localhost`. Instead, add `dataflow.local` to your local DNS hosts file.

resolves #648

### CODE REVIEW

**Comment by ghillert** [comment]
Rebased and squashed commits.

File: ui/protractor-saucelabs.conf.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+const { SpecReporter } = require('jasmine-spec-reporter');
+
+exports.config = {
+  jasmineNodeOpts: {
+    defaultTimeoutInterval: 5000000
+  },
+  sauceUser: process.env.SAUCE_USERNAME,
+  sauceKey: process.env.SAUCE_ACCESS_KEY,
+  multiCapabilities: [
```
Comment (oodamien) [inline ui/protractor-saucelabs.conf.js]: We should define what are the default targets (IE/Safari/Chrome/... and the OS).

File: ui/protractor-saucelabs.conf.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  }
+};
+
+function startSauceConnect(deferred) {
+  var sauceConnectLauncher = require('sauce-connect-launcher');
+  console.log('Launching Sauce Connect...')
+  sauceConnectLauncher(
+    {
+      username: 'ghillert',
+      accessKey: 'fcb9ade1-a0bc-4d7f-9783-e96f067318cb',
```
Comment (oodamien) [inline ui/protractor-saucelabs.conf.js]: Visible access key?

File: ui/protractor-saucelabs.conf.js (hunk 1 above)
Comment (ghillert) [reply ui/protractor-saucelabs.conf.js]: I created a follow-up issue for @sabbyanandan to define:

https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/720

**Pull request merged**